lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3769,6 +3772,110 @@ static bool isReadOnlyResSubscriptOrLoad(CallInst *PtrCI) {
+      collectAllStores(GEP, Stores);
+    } else if (const StoreInst *SI = dyn_cast<StoreInst>(U)) {
+      Stores.emplace_back(SI);
+    } else if (const CallInst *CI = dyn_cast<CallInst>(U)) {
+      Function *F = CI->getCalledFunction();
+      if (F->isIntrinsic()) {
+        if (F->getIntrinsicID() == Intrinsic::lifetime_start ||
+            F->getIntrinsicID() == Intrinsic::lifetime_end)
+          continue;
+      }
+
+      HLOpcodeGroup group = hlsl::GetHLOpcodeGroupByName(F);
+      switch (group) {
+      case HLOpcodeGroup::HLMatLoadStore: {
+        HLMatLoadStoreOpcode opcode =
+            static_cast<HLMatLoadStoreOpcode>(hlsl::GetHLOpcode(CI));
+        switch (opcode) {
+        case HLMatLoadStoreOpcode::ColMatLoad:
+        case HLMatLoadStoreOpcode::RowMatLoad:
+          break;
+        case HLMatLoadStoreOpcode::ColMatStore:
+        case HLMatLoadStoreOpcode::RowMatStore:
+          Stores.emplace_back(CI);
+          break;
+        default:
+          DXASSERT(0, "invalid opcode");
+          Stores.emplace_back(CI);
+          break;
+        }
+      } break;
+      case HLOpcodeGroup::HLSubscript: {
+        HLSubscriptOpcode opcode =
+            static_cast<HLSubscriptOpcode>(hlsl::GetHLOpcode(CI));
+        switch (opcode) {
+        case HLSubscriptOpcode::VectorSubscript:
+        case HLSubscriptOpcode::ColMatElement:
+        case HLSubscriptOpcode::ColMatSubscript:
+        case HLSubscriptOpcode::RowMatElement:
+        case HLSubscriptOpcode::RowMatSubscript:
+          collectAllStores(CI, Stores);
+          break;
+        default:
+          // Rest are resource ptr like buf[i].
+          // Only read of resource handle.
+          break;
+        }
+      } break;
+      default: {
+        // If not sure its out param or not. Take as out param.
+        Stores.emplace_back(CI);
+      }
+      }
+    }
+  }
+}
+
+// return true if A is before B.
+// When A and B in same basic block, before means A will be reach first when
+// iterate the basic block. When A and B is not in same basic block, before
+// means block A will be reach first when doing reverse post order traversal on
+// basic blocks.
+static bool isBefore(const Instruction *A, const Instruction *B,
+                     DenseMap<const BasicBlock *, unsigned> &RPOTOrder) {
+  const BasicBlock *BBA = A->getParent();
+  const BasicBlock *BBB = B->getParent();
+  if (BBA != BBB) {
+    return RPOTOrder[BBA] < RPOTOrder[BBB];
+  }
+  for (const Instruction &I : *BBA) {
+    if (A == &I)
+      return true;
+    if (B == &I)
+      return false;
+  }
+  llvm_unreachable("cannot find A and B in the block");
+  return false;
+}
+
+// Make sure all store on V are before I.
+static bool noUnusedStoreAfter(Value *V, Instruction *I) {
+  SmallVector<const Instruction *, 4> Stores;
+  collectAllStores(V, Stores);
+
+  ReversePostOrderTraversal<Function *> RPOT(I->getParent()->getParent());
```

Comment:
Is this redundant/expensive when we should have a DominatorTree already?

Reply:
Can we at least cache this for the function like the dom tree when it's necessary?

Reply:
I'm not sure :(
This is a rare case,

Reply:
I didn't mean to pre-build this, but to cache it for the functions that required it, so it won't be redundantly computed if called again on the same function.

Reply:
I see. I'll add cache then.

We pre-build the dom tree for all functions. :)

Reply:
Done.